src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -117,18 +118,29 @@ private T runWithRetries(byte[] key, int redirections, boolean tryRandomNode, bo
       }
 
       return execute(connection);
-    } catch (JedisConnectionException jce) {
-      if (tryRandomNode) {
-        // maybe all connection is down
-        throw jce;
-      }
 
+    } catch (JedisNoReachableClusterNodeException jnrcne) {
+      //Not sure if JedisNoReachableClusterNodeException exception should extends JedisConnectionException, it's not connection exception
+      //Remove this if hierarchy reworked and no need to make exception from JedisConnectionException clause
+      throw jnrcne;
+    } catch (JedisConnectionException jce) {
       // release current connection before recursion
       releaseConnection(connection);
       connection = null;
 
-      // retry with random connection
-      return runWithRetries(key, redirections - 1, true, asking);
+      if (redirections <= 1) {
+        //We need this because if node is not reachable anymore - we need to finally initiate slots renewing,
+        //or we can stuck with cluster state without one node in opposite case.
+        //But now if maxRedirections = 1 or 2 we will do it too often. For each time-outed request.
```

Comment:
I'm totally fine to implement more complex solution in one more PR right after this that will avoid rediscovering often than once per time. And it will be correct and fine for your case. And will not make rediscovering if nodes are just unstable, but available. Just need approving of this idea by @marcosnils and @HeartSaVioR in that case.